### PR TITLE
Add documentation for Monadic Comprehensions

### PR DESCRIPTION
**Pull request by pakoito**
Took a bit to get out, I think it turned out okay.

### CODE REVIEW

**Comment by pakoito** [comment]
Missing: links to other typeclasses. Will do in a minute.

**Comment by pakoito** [comment]
Ready for review!

File: kategory-docs/docs/docs/patterns/monadcomprehensions/README.md [hunk 1]
```diff
@@ -6,4 +6,211 @@ permalink: /docs/patterns/monad_comprehensions/
...
+```kotlin:ank
+import kategory.*
+import kategory.effects.*
+
+IO.monad().binding {
+  yields(1)
+}.ev().unsafeRunSync()
+```
+
+Anything in the function inside `binding` can be imperative and sequential code that'll be executed when the datatype decides.
```
Comment (nomisRev) [inline kategory-docs/docs/docs/patterns/monadcomprehensions/README.md]: Isn't data type 2 words?

File: kategory-docs/docs/docs/patterns/monadcomprehensions/README.md [hunk 2]
```diff
@@ -6,4 +6,211 @@ permalink: /docs/patterns/monad_comprehensions/
...
+  IO.monadError().bindingE {
+    val file = bindIn(ioThreadContext) { getFile(path) }
+    val lines = bindIn(computationThreadContext) { file.readLines() }
+    val count = lines.map { it.length }.foldL(0) { acc, lineLength -> acc + lineLength }
+    val average = count / lines.length
+    yields(average)
+  }
+```
+
+Note that `bindIn()` doesn't assure that the execution will return to the same thread where the binding started, as it depends on the implementation of the datatype.
```
Comment (nomisRev) [inline kategory-docs/docs/docs/patterns/monadcomprehensions/README.md]: data type

File: kategory-docs/docs/docs/patterns/monadcomprehensions/README.md [hunk 3]
```diff
@@ -6,4 +6,211 @@ permalink: /docs/patterns/monad_comprehensions/
...
+    val file = getFile(path).bind()
+    val lines = file.readLines().bind()
+    val count = lines.map { it.length }.foldL(0) { acc, lineLength -> acc + lineLength }
+    val average = count / lines.length
+    yields(average)
+  }
+```
+
+With a small change we get handling of exceptions even within the binding block.
+This wrapping works the same way as if we raised an error return from `getFile()` or `readLines()`, shortcircuiting and stopping the sequence early.
```
Comment (nomisRev) [inline kategory-docs/docs/docs/patterns/monadcomprehensions/README.md]: short circuiting and stopping the sequence early.

File: kategory-docs/docs/docs/patterns/monadcomprehensions/README.md [hunk 4]
```diff
@@ -6,4 +6,211 @@ permalink: /docs/patterns/monad_comprehensions/
...
+      IO.raiseError(RuntimeException("File has fewer lines than expected"))
+    } else {
+      yields(lines.take(count))
+    }
+  }
+```
+
+While this looks like a great improvement to manually raise errors sometimes you will encounter unexpected behavior and exceptions in seemingly normal code.
+
+### Error handling in comprehensions
```
Comment (nomisRev) [inline kategory-docs/docs/docs/patterns/monadcomprehensions/README.md]: Perhaps link to `docs/patterns/error_handling`?

Comment (pakoito) [reply kategory-docs/docs/docs/patterns/monadcomprehensions/README.md]: I should rename it, this is about wrapping arbitrary errors in the chain.

**Pull request merged**